File: dotnet/src/IForevents/Integration.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace IForevents
{
    /// <summary>Implemented by every destination. Derive from <see cref="Integration"/> for defaults and hooks.</summary>
    public interface IIntegration
    {
        string Name { get; }
        Task InitAsync(CancellationToken cancellationToken = default);
        Task IdentifyAsync(IdentifyEvent evt, CancellationToken cancellationToken = default);
        Task TrackAsync(TrackEvent evt, CancellationToken cancellationToken = default);
        Task PageAsync(PageEvent evt, CancellationToken cancellationToken = default);
        Task ResetAsync(CancellationToken cancellationToken = default);
        /// <summary>Sends anything buffered.</summary>
        Task FlushAsync(CancellationToken cancellationToken = default);
        /// <summary>Flushes and releases resources.</summary>
        Task ShutdownAsync(CancellationToken cancellationToken = default);
    }

    /// <summary>Optional callbacks fired before an integration handles a call.</summary>
    public sealed class Hooks
    {
        public Action? OnInit { get; set; }
        public Action<IdentifyEvent>? OnIdentify { get; set; }
        public Action<TrackEvent>? OnTrack { get; set; }
        public Action<PageEvent>? OnPage { get; set; }
        public Action? OnReset { get; set; }
    }

    /// <summary>Defaults and hooks; mirrors the Flutter Integration base. Call base first in overrides.</summary>
    public abstract class Integration : IIntegration
    {
        protected Integration(string? name = null, Hooks? hooks = null)
        {
            Name = name ?? GetType().Name;
            HooksConfig = hooks ?? new Hooks();
        }

        public string Name { get; }
        protected Hooks HooksConfig { get; }

        public virtual Task InitAsync(CancellationToken cancellationToken = default)
        {
            HooksConfig.OnInit?.Invoke();
            return Task.CompletedTask;
        }

        public virtual Task IdentifyAsync(IdentifyEvent evt, CancellationToken cancellationToken = default)
        {
            HooksConfig.OnIdentify?.Invoke(evt);
            return Task.CompletedTask;
        }

        public virtual Task TrackAsync(TrackEvent evt, CancellationToken cancellationToken = default)
        {
            HooksConfig.OnTrack?.Invoke(evt);
            return Task.CompletedTask;
        }

        public virtual Task PageAsync(PageEvent evt, CancellationToken cancellationToken = default)
        {
            HooksConfig.OnPage?.Invoke(evt);
            return Task.CompletedTask;
        }

        public virtual Task ResetAsync(CancellationToken cancellationToken = default)
        {
            HooksConfig.OnReset?.Invoke();
            return Task.CompletedTask;
        }

        public virtual Task FlushAsync(CancellationToken cancellationToken = default) => Task.CompletedTask;

        public virtual Task ShutdownAsync(CancellationToken cancellationToken = default) => Task.CompletedTask;

        /// <summary>Runs one integration call in isolation and reports the outcome.</summary>
        public static async Task<IntegrationResult> SafeExecuteAsync(IIntegration integration, Func<Task> action)
        {
            try
            {
                await action().ConfigureAwait(false);
                return new IntegrationResult(integration.Name, true);
            }
            catch (Exception e)
            {
                return new IntegrationResult(integration.Name, false, e);
            }
        }
    }
}

File: go/integration.go
package iforevents

import (
	"context"
	"fmt"
	"time"
)

// Integration is implemented by every destination. Embed BaseIntegration to
// get no-op defaults and hooks, then override what the vendor supports.
type Integration interface {
	Name() string
	Init(ctx context.Context) error
	Identify(ctx context.Context, event IdentifyEvent) error
	Track(ctx context.Context, event TrackEvent) error
	Page(ctx context.Context, event PageEvent) error
	Reset(ctx context.Context) error
	Flush(ctx context.Context) error
	Shutdown(ctx context.Context) error
}

// Hooks are optional callbacks fired before an integration handles a call.
type Hooks struct {
	OnInit     func()
	OnIdentify func(IdentifyEvent)
	OnTrack    func(TrackEvent)
	OnPage     func(PageEvent)
	OnReset    func()
}

// BaseIntegration provides the defaults; mirrors the Flutter Integration base.
type BaseIntegration struct {
	IntegrationName string
	Hooks           Hooks
}

func (b *BaseIntegration) Name() string {
	if b.IntegrationName == "" {
		return "Integration"
	}
	return b.IntegrationName
}

func (b *BaseIntegration) Init(context.Context) error {
	if b.Hooks.OnInit != nil {
		b.Hooks.OnInit()
	}
	return nil
}

func (b *BaseIntegration) Identify(_ context.Context, e IdentifyEvent) error {
	if b.Hooks.OnIdentify != nil {
		b.Hooks.OnIdentify(e)
	}
	return nil
}

func (b *BaseIntegration) Track(_ context.Context, e TrackEvent) error {
	if b.Hooks.OnTrack != nil {
		b.Hooks.OnTrack(e)
	}
	return nil
}

func (b *BaseIntegration) Page(_ context.Context, e PageEvent) error {
	if b.Hooks.OnPage != nil {
		b.Hooks.OnPage(e)
	}
	return nil
}

func (b *BaseIntegration) Reset(context.Context) error {
	if b.Hooks.OnReset != nil {
		b.Hooks.OnReset()
	}
	return nil
}

func (b *BaseIntegration) Flush(context.Context) error    { return nil }
func (b *BaseIntegration) Shutdown(context.Context) error { return nil }

// SafeExecute runs one integration call in isolation (panics included) and
// reports the outcome.
func SafeExecute(integration Integration, action func() error) (result IntegrationResult) {
	result = IntegrationResult{Integration: integration.Name(), Timestamp: time.Now()}
	defer func() {
		if r := recover(); r != nil {
			result.Success = false
			result.Err = fmt.Errorf("panic: %v", r)
		}
	}()
	if err := action(); err != nil {
		result.Err = err
		return result
	}
	result.Success = true
	return result
}
